src/sundials/sundials_hashmap.c

```diff
@@ -74,14 +55,14 @@ struct _SUNHashMap
   **Returns:**
     * A SUNErrCode indicating success or a failure
  */
-static SUNErrCode SUNHashMap_New(int max_size, SUNHashMap* map)
+SUNErrCode SUNHashMap_New(int max_size, SUNHashMap* map)
 {
   int i;
 
   if (max_size <= 0) { return SUN_ERR_ARG_OUTOFRANGE; }
 
   *map = NULL;
-  *map = (SUNHashMap)malloc(sizeof(struct _SUNHashMap));
+  *map = (SUNHashMap)malloc(sizeof(**map));
```

Comment:
Personally, I find this to be less readable in this case.  I like to do `sizeof(type)` in these contexts because IMO it more clearly expresses the intent to allocate memory for `type`.

Reply:
The `sizeof(type)` can be fragile since it's easy to forget to change that if the type of `map` is changed. Also it makes more work when renaming a type like this. https://stackoverflow.com/q/17258647

Reply:
Yeah, in general, but why would we ever change the type of `SUNHashMap` in the malloc constructor for `SUNHashMap`?

Reply:
Generally, I understand the point that `sizeof(expr)` would be less fragile to a chnage of the type, but that does not seem likely to ever happen here.

Reply:
It looks like the rest of sundials is a bit of a mixed bag on this stylistic point, so there isnt really a precedent.

Reply:
> Yeah, in general, but why would we ever change the type of SUNHashMap in the malloc inside the constructor for SUNHashMap?

Yeah, I can't imagine that here. Certainly other places, types could change and cause issues, e.g., switching `long int` to `sunindextype`.

> Generally, I understand the point that sizeof(expr) would be less fragile to a chnage of the type, but that does not seem likely to ever happen here. Furthermore, if it did happen, we would also have to still update the cast.

Yes, but the cast is unnecessary and can be removed.

> It looks like the rest of sundials is a bit of a mixed bag on this stylistic point, so there isnt really a precedent.

I'll go with `sizeof(type)` for now, and we can discuss how to do this more consistently at our next meeting. My 2 cents is we should use `data = malloc(sizeof(*data))` everywhere.

Reply:
In the past we've had some users that pull in the sundials source and build it as part of their C++ application with a C++ compiler and while we don't endorse this use case the cast is required there.

Reply:
Yeah, I think the cast needs to stay for that reason. As far as the `sizeof`, it doesn't really matter but perhaps we can choose a convention to stick with.

Reply:
After thinking about it a bit more, always using `data = malloc(sizeof(*data))` might make any future scripted change to use an allocator easier.

Reply:
Scripting-wise either style is fine as `data = malloc(sizeof(...)` will become something like `sunMalloc(sizeof(...))`

Reply:
I slightly prefer the more verbose style
```
*map = (SUNHashMap)malloc(sizeof(struct SUNHashMap_));
```
to the alternative
```
*map = (SUNHashMap)malloc(sizeof(**map));
```
however now that C99 is allowed, something like
```
SUNHashMap map_out = (SUNHashMap)malloc(sizeof(*map));
```
could also work.

Reply:
Ok (2) it is. When we upgrade to `sunMalloc` we could do something like this to avoid casts

```
#ifdef __cplusplus
template<class T>
T* sunMalloc(size_t size, T* data) { return (T*) malloc(size * sizeof(T)); }
#else
void *sunMalloc(size_t size, void* data) { return malloc(size * sizeof(*data)); }
#endif

sunrealtype *data = sunMalloc(10, data);
```

Reply:
Hmm, `sizeof(*data)` is not right, but the point stands that we can probably split up the C/C++ implementation to avoid casts.

Reply:
We don't explicitly support building with a C++ compiler so I don't think we should worry about splitting this up into C and C++ versions.